src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -325,6 +356,459 @@ CreateInsertSelectRouterPlan(Query *originalQuery,
+		{
+			RestrictInfo *rinfo = (RestrictInfo *) lfirst(restrictionInfoList);
+			OpExpr *restrictionOpExpr = NULL;
+			Expr *leftExpr = NULL;
+			Expr *rightExpr = NULL;
+			Var *leftVar = NULL;
+			Var *rightVar = NULL;
+			Expr *restrictionClause = rinfo->clause;
+
+			if (!IsA(restrictionClause, OpExpr))
+			{
+				continue;
+			}
+
+			restrictionOpExpr = (OpExpr *) restrictionClause;
+			if (list_length(restrictionOpExpr->args) != 2)
+			{
+				continue;
+			}
+			if (!OperatorImplementsEquality(restrictionOpExpr->opno))
+			{
+				continue;
+			}
+
+			leftExpr = linitial(restrictionOpExpr->args);
+			rightExpr = lsecond(restrictionOpExpr->args);
+			if (!(IsA(leftExpr, Var) && IsA(rightExpr, Var)))
+			{
+				continue;
+			}
+
+			leftVar = (Var *) leftExpr;
+			rightVar = (Var *) rightExpr;
+
+			/* we could (probably) safely do that since we disallowed ANTI JOINs ? */
+			AddToVarEquivalenceClass(joinRestriction->plannerInfo, leftVar,
+									 &varEquivalance);
+			AddToVarEquivalenceClass(joinRestriction->plannerInfo, rightVar,
+									 &varEquivalance);
+
+			varEquivalences = lappend(varEquivalences, varEquivalance);
+		}
+	}
+
+	return varEquivalences;
+}
+
+
+/*
+ * ListConcatUniqueEquivalancePartitionKeys get the common equivalence class and
+ * a new equivalence class. It iterates on the members of the new class and adds
+ * the partition key members which are equivalant to the common class.
+ */
+static void
+ListConcatUniqueEquivalantPartitionKeys(VarEquivalenceClass **commonEqClass,
+										List *newEqClass)
+{
+	ListCell *newEqClassCell = NULL;
+
+	foreach(newEqClassCell, newEqClass)
+	{
+		VarEquivalenceClassMember *newEqMember = lfirst(newEqClassCell);
+		Var *relationPartitionKey = PartitionKey(newEqMember->relationId);
+
+		if (!relationPartitionKey)
+		{
+			continue;
+		}
+
+		if (relationPartitionKey->varattno != newEqMember->varattno)
+		{
+			continue;
+		}
+
+		if (VarClassMemberEqualsToVarClass(newEqMember, *commonEqClass))
+		{
+			continue;
+		}
+
+		(*commonEqClass)->equivalentVars = lappend((*commonEqClass)->equivalentVars,
+												   newEqMember);
+	}
+}
+
+
+/*
+ * TODO: (Andres asked) could the rte_relation that we're adding the eq.
+ * be part of ANTI-JOIN?
+ *
+ * AddToVarEquivalenceClass add the varToBeAdded to the given varEquivalanceClass.
+ * The function add the varToBeAdd if it belongs to a relation. If the varToBeAdded
+ * belongs to a subquery, the function recursively calls itself until it finds the
+ * relation that the var actually belongs to.
+ */
+static void
+AddToVarEquivalenceClass(PlannerInfo *root, Var *varToBeAdded,
+						 VarEquivalenceClass **varEquivalanceClass)
+{
+	RangeTblEntry *rte = root->simple_rte_array[varToBeAdded->varno];
+
+	if (rte->rtekind == RTE_RELATION)
+	{
+		VarEquivalenceClassMember *varEqMember = palloc0(
+			sizeof(VarEquivalenceClassMember));
+		varEqMember->varattno = varToBeAdded->varattno;
+		varEqMember->varno = varToBeAdded->varno;
```

Comment:
varno is meaningless outside the Query it was in.  Need to check how it used afterwards.